jep/0000/README.adoc

```diff
@@ -0,0 +1,283 @@
+without adapting the model too.
+* **Heterogeneous data format**: there is no Java API to guide developers on what to return, so the current public CLI and HTTP API is a
+mix of human readable format and XML of all colors.
+* **Outdated serialization tooling**: model is serialized by Stapler and based on old and limited Stapler annotations.
+* **Outdated REST API tooling**: The HTTP API is based on `hudson.model.Api` class which makes nearly impossible to write a proper REST API.
+
+== Reasoning
+
+[TIP]
+====
+Explain why particular design decisions were made.
+Describe alternate designs that were considered and related work, e.g. how the feature is supported in other systems.
+Provide evidence of consensus within the community and discuss important objections or concerns raised during discussion.
+====
+
+There are two main design decisions here:
+
+. **The use of a resources layer instead of relying on model objects directly**. The alternative is "status quo", keep exposing the model through Stapler
+and `doXX` methods (or `getDynamic` to be able to have path parameters).
+There could be an objection on the proposed resources layer: there is more code to write to expose data and sometimes resource classes will
+just mirror the model.
+It is true, but in exchange there is a huge gain on flexibility and maintainability (the model can be modified freely whilst keeping the external data
+API compatible).
+. **Modernizing the tooling to write Data APIs** (introducing SparkJava and Jackson to expose the data). SparkJava is easy to use, simple, lightweight
+and it evolves in a backward compatible way (they say in the documentation). Jackson is the de-facto standard tool to serialize/deserialize JSON nowadays.
+Jackson and SparkJava would be the replacement for Stapler when writing HTTP APIs.
+. **Why Spark?**. There are a lot of frameworks/libraries to create HTTP web services, however, for the goal of this proposal, a lightweight framework
+seems to fit better than others (thinking about Jersey, for example):
+.. SparkJava does not do any classpath scanning to build up the routes, so it works isolatedly based on a series of static calls from plugins code.
+This will avoid potential issues with the "not-so-standard" Jenkins class loader model (and the `JenkinsRule` way of testing).
+.. SparkJava only needs an additional servlet filter to get into the play (which can be defined using already existent Jenkins extension points, see
+https://github.com/amuniz/credentials-plugin/blob/data-api/src/main/java/com/cloudbees/plugins/credentials/api/endpoint/SparkFilter.java[SparkFilter.java]
+which is the only code needed to get SparkJava working in Jenkins).
+.. Docs explicitly say "Spark is always backwards compatible", which is quite important if we build an API on top of it.
+.. Routes are defined statically on startup (as opposite to annotations-based), so it allows the new API to easily perform checks on possible
```

Comment:
What implications will this have for Jenkins and installing new plugins without a restart?

Reply:
Coming from Spring-mvc, I find annotation-based path-routing very easy to understand and handle. But maybe it's just a perception change that I have to do rather than any underlying problem with the technology.

Reply:
> What implications will this have for Jenkins and installing new plugins without a restart?

Extension would need a specific re-evaluation when a new plugin is installed (it already happens with other extensions in Jenkins).

> I find annotation-based path-routing very easy to understand and handle

Yeah, but I wanted a lightweight framework (no classloader introspection, so quick start up time). Also, using the static way to define the routes avoid the temptation to add the `@GET` annotation in the model object directly.